From b279f21a2aef0e892ff97f180eda7f9bfd2de493 Mon Sep 17 00:00:00 2001
From: Dongli Zhang <dongli.zhang0129@gmail.com>
Date: Mon, 22 Nov 2021 07:45:57 -0800
Subject: [PATCH 1/1] libvirt-5.7.0-27

libvirt-5.7.0-27

Signed-off-by: Dongli Zhang <dongli.zhang0129@gmail.com>
---
 src/qemu/qemu_hotplug.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/src/qemu/qemu_hotplug.c b/src/qemu/qemu_hotplug.c
index 09e46113df..aa0e1a627a 100644
--- a/src/qemu/qemu_hotplug.c
+++ b/src/qemu/qemu_hotplug.c
@@ -5130,6 +5130,23 @@ qemuDomainResetDeviceRemoval(virDomainObjPtr vm)
  *      - DEVICE_DELETED event arrived before the timeout time
  *      - we failed to reliably wait for the event and thus use fallback behavior
  */
+/*
+ * (gdb) bt
+ * #0  0x00007efd726aa8b0 in qemuDomainWaitForDeviceRemoval () from /usr/lib64/libvirt/connection-driver/libvirt_driver_qemu.so
+ * #1  0x00007efd726abd6e in qemuDomainSetVcpusLive () from /usr/lib64/libvirt/connection-driver/libvirt_driver_qemu.so
+ * #2  0x00007efd726b7b51 in qemuDomainSetVcpusInternal () from /usr/lib64/libvirt/connection-driver/libvirt_driver_qemu.so
+ * #3  0x00007efd7271c024 in qemuDomainSetVcpusFlags () from /usr/lib64/libvirt/connection-driver/libvirt_driver_qemu.so
+ * #4  0x00007efd9ec8f602 in virDomainSetVcpus () from /lib64/libvirt.so.0
+ * #5  0x0000557863a1ce52 in remoteDispatchDomainSetVcpusHelper ()
+ * #6  0x00007efd9ebb1165 in virNetServerProgramDispatch () from /lib64/libvirt.so.0
+ * #7  0x00007efd9ebb65bd in virNetServerHandleJob () from /lib64/libvirt.so.0
+ * #8  0x00007efd9ead3bd1 in virThreadPoolWorker () from /lib64/libvirt.so.0
+ * #9  0x00007efd9ead2f58 in virThreadHelper () from /lib64/libvirt.so.0
+ * #10 0x00007efd9c0acea5 in start_thread () from /lib64/libpthread.so.0
+ * #11 0x00007efd9bdd59fd in clone () from /lib64/libc.so.6
+ *
+ * virDomainSetVcpus()只被调用一次
+ */
 static int
 qemuDomainWaitForDeviceRemoval(virDomainObjPtr vm)
 {
-- 
2.17.1

