From 7a15387a7935aee8f5860e31d79872016ae58741 Mon Sep 17 00:00:00 2001
From: Dongli Zhang <dongli.zhang0129@gmail.com>
Date: Tue, 22 Dec 2020 11:11:07 -0800
Subject: [PATCH 1/1] linux v5.10

Signed-off-by: Dongli Zhang <dongli.zhang0129@gmail.com>
---
 drivers/xen/mcelog.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/drivers/xen/mcelog.c b/drivers/xen/mcelog.c
index e9ac3b8c4167..1afc5d4a31ed 100644
--- a/drivers/xen/mcelog.c
+++ b/drivers/xen/mcelog.c
@@ -292,6 +292,11 @@ static int convert_log(struct mc_info *mi)
 	return 0;
 }
 
+/*
+ * called by:
+ *   - drivers/xen/mcelog.c|339| <<xen_mce_work_fn>> err = mc_queue_handle(XEN_MC_URGENT);
+ *   - drivers/xen/mcelog.c|344| <<xen_mce_work_fn>> err = mc_queue_handle(XEN_MC_NONURGENT);
+ */
 static int mc_queue_handle(uint32_t flags)
 {
 	struct xen_mc mc_op;
@@ -329,6 +334,13 @@ static int mc_queue_handle(uint32_t flags)
 }
 
 /* virq handler for machine check error info*/
+/*
+ * 在以下调用xen_mce_work=xen_mce_work_fn():
+ *   - drivers/xen/mcelog.c|357| <<xen_mce_interrupt>> schedule_work(&xen_mce_work);
+ *
+ * 在以下使用xen_mce_work_fn():
+ *   - drivers/xen/mcelog.c|353| <<global>> static DECLARE_WORK(xen_mce_work, xen_mce_work_fn);
+ */
 static void xen_mce_work_fn(struct work_struct *work)
 {
 	int err;
@@ -354,6 +366,9 @@ static DECLARE_WORK(xen_mce_work, xen_mce_work_fn);
 
 static irqreturn_t xen_mce_interrupt(int irq, void *dev_id)
 {
+	/*
+	 * xen_mce_work_fn()
+	 */
 	schedule_work(&xen_mce_work);
 	return IRQ_HANDLED;
 }
-- 
2.17.1

