From e81bd676b1aca4281878a6020971b03ede446c18 Mon Sep 17 00:00:00 2001
From: Dongli Zhang <dongli.zhang0129@gmail.com>
Date: Fri, 6 Sep 2019 15:58:05 +0800
Subject: [PATCH 1/1] linux uek5 block

v4.14.35-1902.5.0

Signed-off-by: Dongli Zhang <dongli.zhang0129@gmail.com>
---
 block/blk-mq.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/block/blk-mq.c b/block/blk-mq.c
index ea2d7f9..8760b33 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -2198,6 +2198,11 @@ static void blk_mq_free_map_and_requests(struct blk_mq_tag_set *set,
 	}
 }
 
+/*
+ * called by:
+ *   - block/blk-mq.c|2539| <<blk_mq_init_allocated_queue>> blk_mq_map_swqueue(q);
+ *   - block/blk-mq.c|2583| <<blk_mq_queue_reinit>> blk_mq_map_swqueue(q);
+ */
 static void blk_mq_map_swqueue(struct request_queue *q)
 {
 	unsigned int i, hctx_idx;
@@ -2471,6 +2476,11 @@ static void blk_mq_realloc_hw_ctxs(struct blk_mq_tag_set *set,
 	blk_mq_sysfs_register(q);
 }
 
+/*
+ * called by:
+ *   - block/blk-mq.c|2394| <<blk_mq_init_queue>> q = blk_mq_init_allocated_queue(set, uninit_q);
+ *   - drivers/md/dm-rq.c|807| <<dm_mq_init_request_queue>> q = blk_mq_init_allocated_queue(md->tag_set, md->queue);
+ */
 struct request_queue *blk_mq_init_allocated_queue(struct blk_mq_tag_set *set,
 						  struct request_queue *q)
 {
@@ -2567,6 +2577,10 @@ void blk_mq_free_queue(struct request_queue *q)
 }
 
 /* Basically redo blk_mq_init_queue with queue frozen */
+/*
+ * called by:
+ *   - block/blk-mq.c|2821| <<__blk_mq_update_nr_hw_queues>> blk_mq_queue_reinit(q);
+ */
 static void blk_mq_queue_reinit(struct request_queue *q)
 {
 	WARN_ON_ONCE(!atomic_read(&q->mq_freeze_depth));
-- 
2.7.4

