From 4424f9572177e1ea83aa6b1754f043e1b252987c Mon Sep 17 00:00:00 2001
From: Dongli Zhang <dongli.zhang0129@gmail.com>
Date: Tue, 18 Jun 2019 15:39:27 +0800
Subject: [PATCH 1/1] qemu for 4.0.0

Signed-off-by: Dongli Zhang <dongli.zhang0129@gmail.com>
---
 target/i386/cpu.c | 55 +++++++++++++++++++++++++++++++++++++++++++++++++++++++
 target/i386/cpu.h | 20 ++++++++++++++++++++
 2 files changed, 75 insertions(+)

diff --git a/target/i386/cpu.c b/target/i386/cpu.c
index d6bb57d..958e80b 100644
--- a/target/i386/cpu.c
+++ b/target/i386/cpu.c
@@ -71,6 +71,11 @@ struct CPUID2CacheDescriptorInfo {
  * Known CPUID 2 cache descriptors.
  * From Intel SDM Volume 2A, CPUID instruction
  */
+/*
+ * used by:
+ *   - target/i386/cpu.c|213| <<cpuid2_cache_descriptor>> for (i = 0; i < ARRAY_SIZE(cpuid2_cache_descriptors); i++) {
+ *   - target/i386/cpu.c|214| <<cpuid2_cache_descriptor>> struct CPUID2CacheDescriptorInfo *d = &cpuid2_cache_descriptors[i];
+ */
 struct CPUID2CacheDescriptorInfo cpuid2_cache_descriptors[] = {
     [0x06] = { .level = 1, .type = INSTRUCTION_CACHE, .size =   8 * KiB,
                .associativity = 4,  .line_size = 32, },
@@ -202,6 +207,13 @@ struct CPUID2CacheDescriptorInfo cpuid2_cache_descriptors[] = {
  * Return a CPUID 2 cache descriptor for a given cache.
  * If no known descriptor is found, return CACHE_DESCRIPTOR_UNAVAILABLE
  */
+/*
+ * called by:
+ *   - target/i386/cpu.c|4251| <<cpu_x86_cpuid>> *ecx = cpuid2_cache_descriptor(env->cache_info_cpuid2.l3_cache);
+ *   - target/i386/cpu.c|4253| <<cpu_x86_cpuid>> *edx = (cpuid2_cache_descriptor(env->cache_info_cpuid2.l1d_cache) << 16) |
+ *   - target/i386/cpu.c|4254| <<cpu_x86_cpuid>> (cpuid2_cache_descriptor(env->cache_info_cpuid2.l1i_cache) << 8) |
+ *   - target/i386/cpu.c|4255| <<cpu_x86_cpuid>> (cpuid2_cache_descriptor(env->cache_info_cpuid2.l2_cache));
+ */
 static uint8_t cpuid2_cache_descriptor(CPUCacheInfo *cache)
 {
     int i;
@@ -3023,12 +3035,20 @@ static int cpu_x86_fill_model_id(char *str)
     return 0;
 }
 
+/*
+ * used by:
+ *   - target/i386/cpu.c|3042| <<max_x86_cpu_class_init>> dc->props = max_x86_cpu_properties;
+ */
 static Property max_x86_cpu_properties[] = {
     DEFINE_PROP_BOOL("migratable", X86CPU, migratable, true),
+    /* 只在cpu_x86_cpuid()使用 */
     DEFINE_PROP_BOOL("host-cache-info", X86CPU, cache_info_passthrough, false),
     DEFINE_PROP_END_OF_LIST()
 };
 
+/*
+ * TypeInfo max_x86_cpu_type_info.class_init = max_x86_cpu_class_init()
+ */
 static void max_x86_cpu_class_init(ObjectClass *oc, void *data)
 {
     DeviceClass *dc = DEVICE_CLASS(oc);
@@ -3044,6 +3064,9 @@ static void max_x86_cpu_class_init(ObjectClass *oc, void *data)
 
 static void x86_cpu_load_def(X86CPU *cpu, X86CPUDefinition *def, Error **errp);
 
+/*
+ * TypeInfo max_x86_cpu_type_info.instance_init = max_x86_cpu_initfn()
+ */
 static void max_x86_cpu_initfn(Object *obj)
 {
     X86CPU *cpu = X86_CPU(obj);
@@ -3858,6 +3881,10 @@ static void x86_cpu_apply_props(X86CPU *cpu, PropValue *props)
 
 /* Load data from X86CPUDefinition into a X86CPU object
  */
+/*
+ * called by:
+ *   - target/i386/cpu.c|5630| <<x86_cpu_initfn>> x86_cpu_load_def(cpu, xcc->cpu_def, &error_abort);
+ */
 static void x86_cpu_load_def(X86CPU *cpu, X86CPUDefinition *def, Error **errp)
 {
     CPUX86State *env = &cpu->env;
@@ -4172,6 +4199,28 @@ void cpu_clear_apic_feature(CPUX86State *env)
 
 #endif /* !CONFIG_USER_ONLY */
 
+/*
+ * called by:
+ *   - hw/i386/pc.c|1590| <<pc_build_feature_control_file>> cpu_x86_cpuid(env, 1, 0, &unused, &unused, &ecx, &edx);
+ *   - target/i386/fpu_helper.c|1523| <<helper_xsetbv>> cpu_x86_cpuid(env, 0x0d, 0, &ena_lo, &dummy, &dummy, &ena_hi);
+ *   - target/i386/hvf/hvf.c|793| <<hvf_vcpu_exec>> cpu_x86_cpuid(env, rax, rcx, &rax, &rbx, &rcx, &rdx);
+ *   - target/i386/kvm.c|1051| <<kvm_arch_init_vcpu>> cpu_x86_cpuid(env, 0, 0, &limit, &unused, &unused, &unused);
+ *   - target/i386/kvm.c|1068| <<kvm_arch_init_vcpu>> cpu_x86_cpuid(env, i, 0, &c->eax, &c->ebx, &c->ecx, &c->edx);
+ *   - target/i386/kvm.c|1080| <<kvm_arch_init_vcpu>> cpu_x86_cpuid(env, i, 0, &c->eax, &c->ebx, &c->ecx, &c->edx);
+ *   - target/i386/kvm.c|1094| <<kvm_arch_init_vcpu>> cpu_x86_cpuid(env, i, j, &c->eax, &c->ebx, &c->ecx, &c->edx);
+ *   - target/i386/kvm.c|1119| <<kvm_arch_init_vcpu>> cpu_x86_cpuid(env, i, 0, &c->eax, &c->ebx, &c->ecx, &c->edx);
+ *   - target/i386/kvm.c|1132| <<kvm_arch_init_vcpu>> cpu_x86_cpuid(env, i, j, &c->eax, &c->ebx, &c->ecx, &c->edx);
+ *   - target/i386/kvm.c|1139| <<kvm_arch_init_vcpu>> cpu_x86_cpuid(env, i, 0, &c->eax, &c->ebx, &c->ecx, &c->edx);
+ *   - target/i386/kvm.c|1147| <<kvm_arch_init_vcpu>> cpu_x86_cpuid(env, 0x0a, 0, &eax, &unused, &unused, &edx);
+ *   - target/i386/kvm.c|1171| <<kvm_arch_init_vcpu>> cpu_x86_cpuid(env, 0x80000000, 0, &limit, &unused, &unused, &unused);
+ *   - target/i386/kvm.c|1187| <<kvm_arch_init_vcpu>> cpu_x86_cpuid(env, i, j, &c->eax, &c->ebx, &c->ecx, &c->edx);
+ *   - target/i386/kvm.c|1203| <<kvm_arch_init_vcpu>> cpu_x86_cpuid(env, i, 0, &c->eax, &c->ebx, &c->ecx, &c->edx);
+ *   - target/i386/kvm.c|1210| <<kvm_arch_init_vcpu>> cpu_x86_cpuid(env, 0xC0000000, 0, &limit, &unused, &unused, &unused);
+ *   - target/i386/kvm.c|1221| <<kvm_arch_init_vcpu>> cpu_x86_cpuid(env, i, 0, &c->eax, &c->ebx, &c->ecx, &c->edx);
+ *   - target/i386/misc_helper.c|107| <<helper_cpuid>> cpu_x86_cpuid(env, (uint32_t)env->regs[R_EAX], (uint32_t)env->regs[R_ECX],
+ *
+ * 开机后跑lscpu这个函数就没再调用过
+ */
 void cpu_x86_cpuid(CPUX86State *env, uint32_t index, uint32_t count,
                    uint32_t *eax, uint32_t *ebx,
                    uint32_t *ecx, uint32_t *edx)
@@ -5813,6 +5862,12 @@ static Property x86_cpu_properties[] = {
      * lecacy_cache defaults to true unless the CPU model provides its
      * own cache information (see x86_cpu_load_def()).
      */
+    /*
+     * used by:
+     *   - target/i386/cpu.c|5827| <<global>> DEFINE_PROP_BOOL("legacy-cache", X86CPU, legacy_cache, true),
+     *   - target/i386/cpu.c|3897| <<x86_cpu_load_def>> cpu->legacy_cache = !def->cache_info;
+     *   - target/i386/cpu.c|5269| <<x86_cpu_realizefn>> if (!cpu->legacy_cache) {
+     */
     DEFINE_PROP_BOOL("legacy-cache", X86CPU, legacy_cache, true),
 
     /*
diff --git a/target/i386/cpu.h b/target/i386/cpu.h
index 83fb522..a1f7454 100644
--- a/target/i386/cpu.h
+++ b/target/i386/cpu.h
@@ -1308,6 +1308,19 @@ typedef struct CPUX86State {
     uint32_t cpuid_max_level, cpuid_max_xlevel, cpuid_max_xlevel2;
     /* Actual level/xlevel/xlevel2 value: */
     uint32_t cpuid_level, cpuid_xlevel, cpuid_xlevel2;
+    /*
+     * used by:
+     *   - target/i386/cpu.c|3336| <<x86_cpuid_get_vendor>> x86_cpu_vendor_words2str(value, env->cpuid_vendor1, env->cpuid_vendor2,
+     *   - target/i386/cpu.c|3353| <<x86_cpuid_set_vendor>> env->cpuid_vendor1 = 0;
+     *   - target/i386/cpu.c|3357| <<x86_cpuid_set_vendor>> env->cpuid_vendor1 |= ((uint8_t)value[i ]) << (8 * i);
+     *   - target/i386/cpu.c|4254| <<cpu_x86_cpuid>> *ebx = env->cpuid_vendor1;
+     *   - target/i386/cpu.c|4493| <<cpu_x86_cpuid>> *ebx = env->cpuid_vendor1;
+     *   - target/i386/cpu.c|4508| <<cpu_x86_cpuid>> if (env->cpuid_vendor1 != CPUID_VENDOR_INTEL_1 ||
+     *   - target/i386/cpu.c|5170| <<IS_INTEL_CPU>> #define IS_INTEL_CPU(env) ((env)->cpuid_vendor1 == CPUID_VENDOR_INTEL_1 && \
+     *   - target/i386/cpu.c|5173| <<IS_AMD_CPU>> #define IS_AMD_CPU(env) ((env)->cpuid_vendor1 == CPUID_VENDOR_AMD_1 && \
+     *   - target/i386/translate.c|7188| <<disas_insn>> if (CODE64(s) && env->cpuid_vendor1 != CPUID_VENDOR_INTEL_1)
+     *   - target/i386/translate.c|7199| <<disas_insn>> if (CODE64(s) && env->cpuid_vendor1 != CPUID_VENDOR_INTEL_1)
+     */
     uint32_t cpuid_vendor1;
     uint32_t cpuid_vendor2;
     uint32_t cpuid_vendor3;
@@ -1410,6 +1423,7 @@ struct X86CPU {
     bool vmware_cpuid_freq;
 
     /* if true the CPUID code directly forward host cache leaves to the guest */
+    /* 只在cpu_x86_cpuid()使用 */
     bool cache_info_passthrough;
 
     /* if true the CPUID code directly forwards
@@ -1446,6 +1460,12 @@ struct X86CPU {
     /* Compatibility bits for old machine types.
      * If true present the old cache topology information
      */
+    /*
+     * used by:
+     *   - target/i386/cpu.c|5827| <<global>> DEFINE_PROP_BOOL("legacy-cache", X86CPU, legacy_cache, true),
+     *   - target/i386/cpu.c|3897| <<x86_cpu_load_def>> cpu->legacy_cache = !def->cache_info;
+     *   - target/i386/cpu.c|5269| <<x86_cpu_realizefn>> if (!cpu->legacy_cache) {
+     */
     bool legacy_cache;
 
     /* Compatibility bits for old machine types: */
-- 
2.7.4

