From 5f81add4fb608872070222af933471082d614811 Mon Sep 17 00:00:00 2001
From: Dongli Zhang <dongli.zhang@oracle.com>
Date: Tue, 11 Dec 2018 00:19:34 +0800
Subject: [PATCH 1/1] seabios: comment for seabios-1.12.0

Signed-off-by: Dongli Zhang <dongli.zhang0129@gmail.com>
---
 src/fw/pciinit.c |  4 ++++
 src/hw/pci.c     | 14 ++++++++++++++
 2 files changed, 18 insertions(+)

diff --git a/src/fw/pciinit.c b/src/fw/pciinit.c
index c0634bc..aa11de6 100644
--- a/src/fw/pciinit.c
+++ b/src/fw/pciinit.c
@@ -1146,6 +1146,10 @@ static void pci_bios_map_devices(struct pci_bus *busses)
  * Main setup code
  ****************************************************************/
 
+/*
+ * 只被以下调用:
+ *   - src/fw/paravirt.c|165| <<qemu_platform_setup>> pci_setup();
+ */
 void
 pci_setup(void)
 {
diff --git a/src/hw/pci.c b/src/hw/pci.c
index 9855bad..eb54fcb 100644
--- a/src/hw/pci.c
+++ b/src/hw/pci.c
@@ -11,7 +11,13 @@
 #include "util.h" // udelay
 #include "x86.h" // outl
 
+/*
+ * pci用io port写的时候,先往addr寄存器写地址, 再往data寄存器写数据
+ */
+
+/* 在qemu中是MCH_HOST_BRIDGE_CONFIG_ADDR */
 #define PORT_PCI_CMD           0x0cf8
+/* 在qemu中是MCH_HOST_BRIDGE_CONFIG_DATA */
 #define PORT_PCI_DATA          0x0cfc
 
 void pci_config_writel(u16 bdf, u32 addr, u32 val)
@@ -114,6 +120,14 @@ pci_next(int bdf, int bus)
 int
 pci_probe_host(void)
 {
+    /*
+     * q35应该调用的qemu的pci_host_config_write()
+     *
+     * PCI规范使用从0CF8H~0CFFH 这8个I/O地址来访问所有设备的PCI配置空间.
+     * 这8个字节实际上构成了两个32位寄存器:0CF8H寄存器叫做"配置地址寄存器";
+     * 0CFCH叫做"配置数据寄存器".
+     * 当要访问配置空间的寄存器时,先向地址寄存器写上目标地址,然后就可以从数据寄存器中读写数据了.
+     */
     outl(0x80000000, PORT_PCI_CMD);
     if (inl(PORT_PCI_CMD) != 0x80000000) {
         dprintf(1, "Detected non-PCI system\n");
-- 
2.7.4

