From eee454ad6f6707504294239379c4540650be895e Mon Sep 17 00:00:00 2001
From: Dongli Zhang <dongli.zhang0129@gmail.com>
Date: Sun, 26 Jul 2020 06:50:19 -0700
Subject: [PATCH 1/1] sysstat-v10.1.6

Signed-off-by: Dongli Zhang <dongli.zhang0129@gmail.com>
---
 iostat.c | 84 +++++++++++++++++++++++++++++++++++++++++++++++++++++++-
 1 file changed, 83 insertions(+), 1 deletion(-)

diff --git a/iostat.c b/iostat.c
index 171d9ae..5cb343b 100644
--- a/iostat.c
+++ b/iostat.c
@@ -53,6 +53,20 @@ char *sccsid(void) { return (SCCSID); }
 struct stats_cpu *st_cpu[2];
 unsigned long long uptime[2]  = {0, 0};
 unsigned long long uptime0[2] = {0, 0};
+/*
+ * 在以下使用st_iodev[]:
+ *   - iostat.c|213| <<salloc_device>> if ((st_iodev[i] =
+ *   - iostat.c|218| <<salloc_device>> memset(st_iodev[i], 0, IO_STATS_SIZE * dev_nr);
+ *   - iostat.c|411| <<io_sys_free>> if (st_iodev[i]) {
+ *   - iostat.c|412| <<io_sys_free>> free(st_iodev[i]);
+ *   - iostat.c|462| <<save_stats>> st_iodev_i = st_iodev[!curr] + i;
+ *   - iostat.c|473| <<save_stats>> st_iodev_i = st_iodev[curr] + i;
+ *   - iostat.c|794| <<compute_device_groups_stats>> ioi = st_iodev[curr] + i;
+ *   - iostat.c|1140| <<write_stats>> ioi = st_iodev[curr] + i;
+ *   - iostat.c|1141| <<write_stats>> ioj = st_iodev[!curr] + i;
+ *
+ * *st_iodev[2]是两个指针
+ */
 struct io_stats *st_iodev[2];
 struct io_hdr_stats *st_hdr_iodev;
 struct io_dlist *st_dev_list;
@@ -298,6 +312,10 @@ int update_dev_list(int *dlist_idx, char *device_name)
  * Allocate and init structures, according to system state.
  ***************************************************************************
  */
+/*
+ * called by:
+ *   - iostat.c|1647| <<main>> io_sys_init();
+ */
 void io_sys_init(void)
 {
 	/* Allocate and init stat common counters */
@@ -433,6 +451,12 @@ void io_sys_free(void)
  * @st_hdr_iodev	Pointer on structures describing a device/partition.
  ***************************************************************************
  */
+/*
+ * called by:
+ *   - iostat.c|553| <<read_sysfs_file_stat>> save_stats(dev_name, curr, &sdev, iodev_nr, st_hdr_iodev);
+ *   - iostat.c|780| <<read_diskstats_stat>> save_stats(dev_name, curr, &sdev, iodev_nr, st_hdr_iodev);
+ *   - iostat.c|829| <<compute_device_groups_stats>> save_stats(shi->name, curr, &gdev, iodev_nr, st_hdr_iodev);
+ */
 void save_stats(char *name, int curr, void *st_io, int iodev_nr,
 		struct io_hdr_stats *st_hdr_iodev)
 {
@@ -459,6 +483,20 @@ void save_stats(char *name, int curr, void *st_io, int iodev_nr,
 				/* Unused entry found... */
 				st_hdr_iodev_i->used = TRUE; /* Indicate it is now used */
 				strcpy(st_hdr_iodev_i->name, name);
+				/*
+				 * 在以下使用st_iodev[]:
+				 *   - iostat.c|213| <<salloc_device>> if ((st_iodev[i] =
+				 *   - iostat.c|218| <<salloc_device>> memset(st_iodev[i], 0, IO_STATS_SIZE * dev_nr);
+				 *   - iostat.c|411| <<io_sys_free>> if (st_iodev[i]) {
+				 *   - iostat.c|412| <<io_sys_free>> free(st_iodev[i]);
+				 *   - iostat.c|462| <<save_stats>> st_iodev_i = st_iodev[!curr] + i;
+				 *   - iostat.c|473| <<save_stats>> st_iodev_i = st_iodev[curr] + i;
+				 *   - iostat.c|794| <<compute_device_groups_stats>> ioi = st_iodev[curr] + i;
+				 *   - iostat.c|1140| <<write_stats>> ioi = st_iodev[curr] + i;
+				 *   - iostat.c|1141| <<write_stats>> ioj = st_iodev[!curr] + i;
+				 *
+				 * *st_iodev[2]是两个指针
+				 */
 				st_iodev_i = st_iodev[!curr] + i;
 				memset(st_iodev_i, 0, IO_STATS_SIZE);
 				break;
@@ -470,6 +508,20 @@ void save_stats(char *name, int curr, void *st_io, int iodev_nr,
 		if (st_hdr_iodev_i->status == DISK_UNREGISTERED) {
 			st_hdr_iodev_i->status = DISK_REGISTERED;
 		}
+		/*
+		 * 在以下使用st_iodev[]:
+		 *   - iostat.c|213| <<salloc_device>> if ((st_iodev[i] =
+		 *   - iostat.c|218| <<salloc_device>> memset(st_iodev[i], 0, IO_STATS_SIZE * dev_nr);
+		 *   - iostat.c|411| <<io_sys_free>> if (st_iodev[i]) {
+		 *   - iostat.c|412| <<io_sys_free>> free(st_iodev[i]);
+		 *   - iostat.c|462| <<save_stats>> st_iodev_i = st_iodev[!curr] + i;
+		 *   - iostat.c|473| <<save_stats>> st_iodev_i = st_iodev[curr] + i;
+		 *   - iostat.c|794| <<compute_device_groups_stats>> ioi = st_iodev[curr] + i;
+		 *   - iostat.c|1140| <<write_stats>> ioi = st_iodev[curr] + i;
+		 *   - iostat.c|1141| <<write_stats>> ioj = st_iodev[!curr] + i;
+		 *
+		 * *st_iodev[2]是两个指针
+		 */
 		st_iodev_i = st_iodev[curr] + i;
 		*st_iodev_i = *((struct io_stats *) st_io);
 	}
@@ -682,6 +734,11 @@ void read_sysfs_stat(int curr)
  * @curr	Index in array for current sample statistics.
  ***************************************************************************
  */
+/*
+ * called by:
+ *   - iostat.c|1271| <<rw_io_stat_loop>> read_diskstats_stat(curr);
+ *   - iostat.c|1283| <<rw_io_stat_loop>> read_diskstats_stat(curr);
+ */
 void read_diskstats_stat(int curr)
 {
 	FILE *fp;
@@ -696,8 +753,15 @@ void read_diskstats_stat(int curr)
 	unsigned int major, minor;
 
 	/* Every I/O device entry is potentially unregistered */
+	/*
+	 * iodev_nr是全局变量
+	 */
 	set_entries_unregistered(iodev_nr, st_hdr_iodev);
 
+	/*
+	 * 这里直接读"/proc/diskstats"
+	 * 不用之前获得过的iodev_nr
+	 */
 	if ((fp = fopen(DISKSTATS, "r")) == NULL)
 		return;
 
@@ -907,6 +971,10 @@ void write_disk_stat_header(int *fctr)
  * @ioj		Previous sample statistics.
  ***************************************************************************
  */
+/*
+ * called by:
+ *   - iostat.c|1263| <<write_stats>> write_ext_stat(curr, itv, fctr, shi, ioi, ioj);
+ */
 void write_ext_stat(int curr, unsigned long long itv, int fctr,
 		    struct io_hdr_stats *shi, struct io_stats *ioi,
 		    struct io_stats *ioj)
@@ -1136,7 +1204,21 @@ void write_stats(int curr, struct tm *rectime)
 						/* Device not found in list: Don't display it */
 						continue;
 				}
-	
+
+				/*
+				 * 在以下使用st_iodev[]:
+				 *   - iostat.c|213| <<salloc_device>> if ((st_iodev[i] =
+				 *   - iostat.c|218| <<salloc_device>> memset(st_iodev[i], 0, IO_STATS_SIZE * dev_nr);
+				 *   - iostat.c|411| <<io_sys_free>> if (st_iodev[i]) {
+				 *   - iostat.c|412| <<io_sys_free>> free(st_iodev[i]);
+				 *   - iostat.c|462| <<save_stats>> st_iodev_i = st_iodev[!curr] + i;
+				 *   - iostat.c|473| <<save_stats>> st_iodev_i = st_iodev[curr] + i;
+				 *   - iostat.c|794| <<compute_device_groups_stats>> ioi = st_iodev[curr] + i;
+				 *   - iostat.c|1140| <<write_stats>> ioi = st_iodev[curr] + i;
+				 *   - iostat.c|1141| <<write_stats>> ioj = st_iodev[!curr] + i;
+				 *
+				 * *st_iodev[2]是两个指针
+				 */
 				ioi = st_iodev[curr] + i;
 				ioj = st_iodev[!curr] + i;
 
-- 
2.17.1

