From 53bf0baf714d5c9835cf94a5ef83c273128e733b Mon Sep 17 00:00:00 2001
From: Dongli Zhang <dongli.zhang@oracle.com>
Date: Fri, 27 Sep 2019 15:55:52 +0800
Subject: [PATCH 1/1] xen-4.4.4-222-comment

xen-4.4.4-222

Signed-off-by: Dongli Zhang <dongli.zhang@oracle.com>
---
 xen/arch/x86/domain_page.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/xen/arch/x86/domain_page.c b/xen/arch/x86/domain_page.c
index 56f9bb1..6ec449c 100644
--- a/xen/arch/x86/domain_page.c
+++ b/xen/arch/x86/domain_page.c
@@ -172,6 +172,23 @@ void *map_domain_page(unsigned long mfn)
     return (void *)MAPCACHE_VIRT_START + pfn_to_paddr(idx);
 }
 
+/*
+ * MAPCACHE_VCPU_ENTRIES = 0x00000010
+ * MAPCACHE_ENTRIES      = 0x0000000000020000
+ * MAPCACHE_VIRT_START   = 0xffff820040000000
+ * MAPCACHE_VIRT_END     = 0xffff820060000000
+ *
+ * DIRECTMAP_VIRT_START = 0xffff830000000000
+ * DIRECTMAP_SIZE       = 0x00007c8000000000
+ * DIRECTMAP_VIRT_END   = 0xffffff8000000000
+ *
+ * LINEAR_PT_VIRT_START = 0xffff810000000000
+ * LINEAR_PT_VIRT_END   = 0xffff818000000000
+ *
+ * HYPERVISOR_VIRT_START = 0xffff800000000000
+ * HYPERVISOR_VIRT_END   = 0xffff880000000000
+ */
+
 void unmap_domain_page(const void *ptr)
 {
     unsigned int idx;
-- 
2.7.4

